### ISSUE TITLE
Task timestamps deep dive

### ISSUE THREAD

**Issue by hschallhorn** [department-of-veterans-affairs/caseflow#12455]
## OR: What do these timestamps even mean?

Spawned from [this](https://dsva.slack.com/archives/CHX8FMP28/p1571764377224900) slack conversation.
Currently, we have a bunch of timestamps on every task.

1. `created_at`: When the `ActiveRecord` was created (done automatically by rails)
1. `updated_at`: When any field of the `ActiveRecord` was last updated (done automatically by rails)
1. `assigned_at`: When the `ActiveRecord` [is created](https://github.com/department-of-veterans-affairs/caseflow/blob/9b73d5ff8aa6a65be9a529b258bde91caa65de82/app/models/task.rb#L27)it is set to the creation datetime OR whatever `assigned_at` attribute is [passed to the task](https://github.com/department-of-veterans-affairs/caseflow/blob/9b73d5ff8aa6a65be9a529b258bde91caa65de82/app/models/task.rb#L626) upon creation (we'll get back to this later). It is also set when a task's `status` is [updated](https://github.com/department-of-veterans-affairs/caseflow/blob/9b73d5ff8aa6a65be9a529b258bde91caa65de82/app/models/task.rb#L644) to "[assigned](https://github.com/department-of-veterans-affairs/caseflow/blob/9b73d5ff8aa6a65be9a529b258bde91caa65de82/app/models/task.rb#L631)"
1. `started_at`: When a task's `status` is [updated](https://github.com/department-of-veterans-affairs/caseflow/blob/9b73d5ff8aa6a65be9a529b258bde91caa65de82/app/models/task.rb#L644) to "[in_progress](https://github.com/department-of-veterans-affairs/caseflow/blob/9b73d5ff8aa6a65be9a529b258bde91caa65de82/app/models/task.rb#L632)"
1. `placed_on_hold_at`: When a task's `status` is [updated](https://github.com/department-of-veterans-affairs/caseflow/blob/9b73d5ff8aa6a65be9a529b258bde91caa65de82/app/models/task.rb#L644) to "[on_hold](https://github.com/department-of-veterans-affairs/caseflow/blob/9b73d5ff8aa6a65be9a529b258bde91caa65de82/app/models/task.rb#L633)"
1. `closed_at`: When a task's `status` is [updated](https://github.com/department-of-veterans-affairs/caseflow/blob/9b73d5ff8aa6a65be9a529b258bde91caa65de82/app/models/task.rb#L644) to "[completed](https://github.com/department-of-veterans-affairs/caseflow/blob/9b73d5ff8aa6a65be9a529b258bde91caa65de82/app/models/task.rb#L634)" or "[cancelled](https://github.com/department-of-veterans-affairs/caseflow/blob/9b73d5ff8aa6a65be9a529b258bde91caa65de82/app/models/task.rb#L635)"

### The problem
A judge has tasks that showed up in their queue yesterday, but show a days waiting of 45 days. The production data supports this, oddly showing an `assigned_at` date before the `created_at` date. The task tree shows that this case was most likely assigned to another judge and then reassigned to the current judge.
```ruby
Appeal 2809 [id, status, assigned_at, created_at]
└── RootTask 177581, on_hold, 2019-04-18 15:20:45 UTC, 2019-04-18 15:20:45 UTC
    ├── TrackVeteranTask 177582, in_progress, 2019-04-18 15:20:45 UTC, 2019-04-18 15:20:45 UTC
    ├── DistributionTask 177583, completed, 2019-04-18 15:20:45 UTC, 2019-04-18 15:20:45 UTC
    ├── JudgeAssignTask 343360, completed, 2019-08-21 17:08:36 UTC, 2019-08-21 17:08:36 UTC
    ├── JudgeDecisionReviewTask 372135, cancelled, 2019-09-20 14:40:55 UTC, 2019-09-04 12:49:42 UTC
    │   └── AttorneyTask 372136, completed, 2019-09-04 12:49:42 UTC, 2019-09-04 12:49:42 UTC
    ├── JudgeDecisionReviewTask 468272, cancelled, 2019-09-20 14:40:55 UTC, 2019-10-15 16:01:21 UTC
    └── JudgeDecisionReviewTask 477406, assigned, 2019-09-20 14:40:55 UTC, 2019-10-21 14:59:33 UTC
```

### Why this happens
We use the `assigned_at` field to determine the number of days the case has been waiting for the assignee to take action and the `placed_on_hold_at` field to determine the number of days the case has been on hold (waiting for something else to happen or for a down stream user to take action). Caseflow users see "Days waiting" for all of the cases in their assigned tab and use it to determine how long it takes them to take action on a case. When we "reassign" a task, we [duplicate the task](https://github.com/department-of-veterans-affairs/caseflow/blob/9b73d5ff8aa6a65be9a529b258bde91caa65de82/app/models/task.rb#L469) to reassign, assign it to a new person, and cancel the old task. All fields that are not stipulated before the duplicate task is saved are copied over from the old task. This means that all of the timestamps that are set on the old task are replicated in the new task.

To some extent, this makes some sense. If the original task was assigned by a judge a few months ago, the case has been waiting for a few months at the attorney stage. However, if it was reassigned yesterday, it has only been waiting on the current user for a day. What should be be displaying to the user?

Similarly, for on hold tasks, as we will be moving the old task's children [to the new task](https://github.com/department-of-veterans-affairs/caseflow/blob/9b73d5ff8aa6a65be9a529b258bde91caa65de82/app/models/task.rb#L478), should the `placed_on_hold_at` represent when the original task was placed on hold?

In my experience, users use these metrics to report to superiors on how long it takes them to complete a case (which explains this judges concern about a case showing as waiting on them for a month and a half when they have only had the case for a day).

### Acceptance  Criteria
- [ ] Determine what these timestamps should represent to the user and whether they should be preserved from the old task when reassigning
- [ ] Write ticket to change any code based on the outcome of the above

**Comment by pkarman** [comment]
I believe we should set `assigned_at` to `Time.zone.now` and `placed_on_hold_at` to `nil` when we clone tasks. The task should apply to the state of the current assignment, and not carry over old data for those 2 timestamps.

If we need to compute the "Days Waiting" for reporting on case lifetime, we should look at the oldest `assigned_at` value of the relevant task. Maybe for the queue display we need to clarify via help text or a second column that there is a difference between the time the case has been waiting and how long it has been assigned to a particular user.

**Comment by pkarman** [comment]
There are currently 39 assigned tasks fitting this description:
```
JudgeDecisionReviewTask.assigned.where("assigned_at < created_at")
```

**Comment by hschallhorn** [comment]
I agreed with your take on what timestamps to reset. Interesting wrench from an ask from VLJ support:

**Comment by hschallhorn** [comment]
We have also head from supervisors that they would like their team members to see this larger "Days waiting" so they can prioritize cases that have been waiting the longest, even if they have only been with the current user for a day

Perhaps the solution here is to clarify to users what days waiting means. I think we should stress to the users that “Days waiting” does not mean “Days waiting on YOU”, it means “Days waiting for an attorney assignment”, “Days waiting for a decision”, “Days waiting for a translation”, etc.

I'm not sure how each team uses these numbers for reporting though.

**Comment by hschallhorn** [comment]
More context

## What we use these timestamps for
### `assigned_at`
- [ ] [Order decision reviews](https://github.com/department-of-veterans-affairs/caseflow/blob/a0900a333636fb4d4b5318fee5349faafa2ae9dd/app/controllers/decision_reviews_controller.rb#L53) to show to business lines
- [ ] [Determine `ready_for_distribution_at` ](https://github.com/department-of-veterans-affairs/caseflow/blob/dbd12673ccf72aea3b8f51af7e6e5440b273837e/app/models/appeal.rb#L228) (which relies on `assigned_at` being reset after coming back from on hold)
- [ ] Check if automatic case distribution is valid by checking if any judge active tasks have been waiting [longer than 30 days](https://github.com/department-of-veterans-affairs/caseflow/blob/43562ed9e75b14c6f949802521da9df6a4214c2b/app/models/distribution.rb#L97) (not resetting assigned_at would also affect this)
- [ ] Ordering dockets by when a distribution task was marked [ready for distribution](https://github.com/department-of-veterans-affairs/caseflow/blob/c08b7acc77fcdf50d9947e6b6f80bae005bd3a02/app/models/docket.rb#L122)
- [ ] Calculating day waiting on a task (what prompted this deep dive)
- [ ] [Mocking the `placed_on_hold_at` date](https://github.com/department-of-veterans-affairs/caseflow/blob/e1e716f233c638dca81ee3064b3d58a9ccbedba4/app/models/queues/attorney_queue.rb#L17) for legacy colocated tasks in an attorney's on hold tab
- [ ] Determine [which attorney task](https://github.com/department-of-veterans-affairs/caseflow/blob/181719c9021f86c63ce7736db413b05234c3b75d/app/workflows/bulk_task_reassignment.rb#L208) to reassign
- [ ] Displaying [when a decision was submitted](https://github.com/department-of-veterans-affairs/caseflow/blob/0ec98c06bdf3d76d2059f4e94c858b90d3c12eb9/client/app/queue/caseEvaluation/EvaluateDecisionView.jsx#L175) and calculating how many [days the case was worked](https://github.com/department-of-veterans-affairs/caseflow/blob/0ec98c06bdf3d76d2059f4e94c858b90d3c12eb9/client/app/queue/caseEvaluation/EvaluateDecisionView.jsx#L176)
- [ ] [Due date column](https://github.com/department-of-veterans-affairs/caseflow/blob/447ecf676784fc1a220f8a1194115b20d742fc2b/client/app/queue/components/TaskTable.jsx#L124) in queue
- [ ] [Days waiting column](https://github.com/department-of-veterans-affairs/caseflow/blob/447ecf676784fc1a220f8a1194115b20d742fc2b/client/app/queue/components/TaskTableColumns.jsx#L208) in queue

**Comment by hschallhorn** [comment]
### `placed_on_hold_at`
- [ ] Determining if a [task has a completed hold](https://github.com/department-of-veterans-affairs/caseflow/blob/502058571a5d398f413592dfa3e737c2d9ae6515/client/app/queue/utils.js#L537)
- [ ] Showing [number of days a task has been on hold](https://github.com/department-of-veterans-affairs/caseflow/blob/d1c345ae4a84c7ac0f0ef94f87e0397a1c847655/client/app/queue/components/OnHoldLabel.jsx#L6)
- [ ] Determining how to display the [on hold progress bar](https://github.com/department-of-veterans-affairs/caseflow/blob/447ecf676784fc1a220f8a1194115b20d742fc2b/client/app/queue/components/TaskTableColumns.jsx#L217)

**Comment by hschallhorn** [comment]
More user feedback/questions
> Tasks that come off of a hold are resetting to zero days assigned Is this correct?

This is due to us updating a timestamp any time a task status is changed to the timestamps respective status (assigned and assigned_at in this case) https://github.com/department-of-veterans-affairs/caseflow/blob/761e76a628fb5b5a2ea322755d51133d8e33ffe8/app/models/task.rb#L634-L643
